File: 2225.find-players-with-zero-or-one-losses/solution.go
func findWinners(matches [][]int) [][]int {
	n := len(matches)
	
	matchMap := make(map[int][2]int)
	for i := 0; i < n; i++ {
			match := matches[i]
			winner := match[0]
			loser := match[1]
			
			winnerStatsRecords := matchMap[winner]
			winnerStatsRecords[0]++
			matchMap[winner] = winnerStatsRecords
			
			loserStatsRecords := matchMap[loser]
			loserStatsRecords[1]++
			matchMap[loser] = loserStatsRecords
	}
	
	alwaysWinPlayers := make([]int, 0)
	loseExactlyOnePlayers := make([]int, 0)
	for player, stats := range matchMap {
			winCnt := stats[0]
			loseCnt := stats[1]
			
			if winCnt > 0 && loseCnt == 0 {
					alwaysWinPlayers = append(alwaysWinPlayers, player)
			} else if loseCnt == 1 {
					loseExactlyOnePlayers = append(loseExactlyOnePlayers, player)
			}
	}
	
	sort.Ints(alwaysWinPlayers)
	sort.Ints(loseExactlyOnePlayers)
	
	return [][]int{ alwaysWinPlayers, loseExactlyOnePlayers }
}

File: 2225.find-players-with-zero-or-one-losses/solution.kt
class Solution {
    fun findWinners(matches: Array<IntArray>): List<List<Int>> {
        // TreeMap sorts keys automatically
        val losesMap = TreeMap<Int, Int>()
        
        for (match in matches) {
            val winner = match[0]
            val loser = match[1]
            
            losesMap[winner] = losesMap[winner] ?: 0
            losesMap[loser] = (losesMap[loser] ?: 0) + 1
        }
        
        val alwaysWinPlayers = ArrayList<Int>()
        val exactlyOnceLosers = ArrayList<Int>()
        for ((player, loseCnt) in losesMap) {
            if (loseCnt == 0) {
                alwaysWinPlayers.add(player)
                continue
            }
            if (loseCnt == 1) {
                exactlyOnceLosers.add(player)
            }
        }
        return listOf(
            alwaysWinPlayers.toList(),
            exactlyOnceLosers.toList(),
        )
    }
}